### PR TITLE
support whitelist of accounts that will not be remove dust

### PR DESCRIPTION
**Pull request by wangjj9219**

### CODE REVIEW

File: currencies/src/mock.rs [hunk 1]
```diff
@@ -84,6 +91,7 @@ impl orml_tokens::Config for Runtime {
 	type ExistentialDeposits = ExistentialDeposits;
 	type OnDust = orml_tokens::TransferDust<Runtime, DustAccount>;
 	type MaxLocks = MaxLocks;
+	type NeverReapedWhitelist = MockNeverReapedWhitelist;
```
Comment (xlc) [inline currencies/src/mock.rs:94]:
```suggestion
	type NeverReapedWhitelist = ();
```

File: tokens/src/lib.rs [hunk 2]
```diff
@@ -197,6 +197,10 @@ pub mod module {
 		type OnDust: OnDust<Self::AccountId, Self::CurrencyId, Self::Balance>;
 
 		type MaxLocks: Get<u32>;
+
+		// The whitelist of accounts that will not be reaped even if its total
+		// is zero or below ED.
+		type NeverReapedWhitelist: Contains<Self::AccountId>;
```
Comment (shaunxw) [inline tokens/src/lib.rs:203]: The name `DustRemovalWhitelist` might be more consistent.

File: tokens/src/lib.rs [hunk 3]
```diff
@@ -728,8 +737,9 @@ impl<T: Config> Pallet<T> {
 	}
 
 	/// Transfer some free balance from `from` to `to`. Ensure from_account
-	/// allow death or new balance above existential deposit. Ensure to_account
-	/// new balance above existential deposit.
+	/// allow death or new balance above existential deposit, and ensure
+	/// to_account new balance above existential deposit, but except those
+	/// senders/receivers are in the never reaped whitelist.
```
Comment (shaunxw) [inline tokens/src/lib.rs:742]:
```suggestion
	/// senders/receivers are in `T::DustRemovalWhitelist`.
```

File: tokens/src/lib.rs [hunk 4]
```diff
@@ -772,7 +790,12 @@ impl<T: Config> Pallet<T> {
 	}
 
 	/// Withdraw some free balance from an account, respecting existence
-	/// requirements, and if AllowDeath do not require ED.
+	/// requirements except for those accounts are in the never reaped
+	/// whitelist, and if AllowDeath do not require ED
```
Comment (shaunxw) [inline tokens/src/lib.rs:794]:
```suggestion
	/// requirements except for those accounts are in `T::DustRemovalWhitelist`,
	/// and if AllowDeath do not require ED
```

**Review by shaunxw — APPROVED** [review]
LGTM. Some nitpicks.

**Comment by wangjj9219** [comment]
@xlc @shaunxw 
I refactor impl to use DustRemovalWhitelist instead of NeverReapedWhitelist:
if an account is in DustRemovalWhitelist and total is not zero and below ED,  will not remove its dust
if an account is in DustRemovalWhitelist and total is zero and ED is not zero, there's no dust but still reap it.

**Review by xlc — APPROVED** [review]

**Review by shaunxw — APPROVED** [review]

**Pull request merged**